### ISSUE TITLE
HTTP Downloader automatically closes when downloading a large list

### ISSUE THREAD

**Issue by rayman3003** [erickutcher/httpdownloader#98]
When I add this list of 9999 links to the HTTP Downloader, after random number of downloads (one time only 50 downloads, & one time after 2900 downloads, & ...), HTTP Downloader closes itself automatically.

[1.txt](https://github.com/erickutcher/httpdownloader/files/3899103/1.txt)

**Comment by erickutcher** [comment]
Sounds like a crash. What version are you using and is it the 64-bit or 32-bit one?

**Comment by rayman3003** [comment]
> 
> 
> Sounds like a crash. What version are you using and is it the 64-bit or 32-bit one?

The last version x64 version: 1.0.2.9

Yes, Its a crash.

I even tried another list from another website (about 

links) & "HTTP Downloader" crashed there too. It means that crashes r nothing to do with large lists, But its a common & serious crash bug.

The worst part is that after crash, When I open the "HTTP Downloader" again, It won't even save the link lists. (Despite the "Enable Download History" get enabled too)

**Comment by erickutcher** [comment]
The list gets saved after all downloading items finish. A crash would prevent that from occurring.

If you could upload your settings file here as well, that would help me out. I just need to be able to recreate everything to test it out.

**Comment by rayman3003** [comment]
> What happens if you use an older version of the program like this one: https://github.com/erickutcher/httpdownloader/releases/download/v1.0.2.4/HTTP_Downloader_64.zip

It works with this old version, without any problem.

And the settings in both old & new version are the same.

So, probably there is a bug in the newer version.

**Comment by erickutcher** [comment]
I think I figured it out. There's a crash happening when the download history is being saved and is really large. I'll upload a test version for you within the next couple of days.

**Comment by erickutcher** [comment]
Go ahead and try this version out: [HTTP_Downloader_64_1.0.3.0_Beta_1.zip](https://github.com/erickutcher/httpdownloader/files/3917647/HTTP_Downloader_64_1.0.3.0_Beta_1.zip)

You can completely replace the current version with this without any issue.

Test it like you did before to try and get the bug to pop up. Do a bunch of tests just to be sure. Let me know what you find.

**Comment by rayman3003** [comment]
> 
> 
> Go ahead and try this version out: [HTTP_Downloader_64_1.0.3.0_Beta_1.zip](https://github.com/erickutcher/httpdownloader/files/3917647/HTTP_Downloader_64_1.0.3.0_Beta_1.zip)
> 
> You can completely replace the current version with this without any issue.
> 
> Test it like you did before to try and get the bug to pop up. Do a bunch of tests just to be sure. Let me know what you find.

Thank u. But unfortunately this beta version also crashing in the middle of download (after downloading 424 files out of 9999 links (of 1.txt in the first post).

**Comment by erickutcher** [comment]
I need your settings file to recreate things on my end. Also, are you setting the number of parts for each download to something other that 1?

**Comment by erickutcher** [comment]
Give this version a try: [HTTP_Downloader_64_1.0.3.0_Beta_2.zip](https://github.com/erickutcher/httpdownloader/files/3920021/HTTP_Downloader_64_1.0.3.0_Beta_2.zip)

I found a crash related to how it handles the range list when the program retries a failed download.

**Comment by rayman3003** [comment]
> 
> 
> I need your settings file to recreate things on my end. Also, are you setting the number of parts for each download to something other that 1?
> 
> Also, to make it easier for me to figure this out, does the crash happen when you download those links using the "Simulate download" option?
> 
> Let me know your operating system too.

Windows 10 x64 1909

Here is my settings:

![2019-12-04 13_43_06-Options](https://user-images.githubusercontent.com/17022701/70135058-8af24b00-169e-11ea-9f7b-5ecfd45375d1.jpg)

**Comment by rayman3003** [comment]
> 
> 
> Give this version a try: [HTTP_Downloader_64_1.0.3.0_Beta_2.zip](https://github.com/erickutcher/httpdownloader/files/3920021/HTTP_Downloader_64_1.0.3.0_Beta_2.zip)
> 
> I found a crash related to how it handles the range list when the program retries a failed download.

First time I tried downloading with this Beta2 version, it crashed just in the beginning of the download (9999 links).

But the second time that I opened this Beta2 version & start the download, It finished all the 9999 links download, without any crash.

So it seems the crash has been solved, somehow!

**Comment by erickutcher** [comment]
Do you split the downloads into parts? In other words, in the Add URL(s) window's Advanced options, do you set the "Download parts" to something greater than 1?

Keep testing things if you don't mind. I gotta know for sure if it's been fixed.

**Comment by smaragdus** [comment]
@rayman3003 

Wouldn't it be better to upload the settings file ('http_downloader_settings') (in ZIP archive) as I suppose it would facilitate testing on developer's part.

**Comment by rayman3003** [comment]
> 
> 
> Do you split the downloads into parts? In other words, in the Add URL(s) window's Advanced options, do you set the "Download parts" to something greater than 1?
> 
> Keep testing things if you don't mind. I gotta know for sure if it's been fixed.

Download parts is 1.

OK, I will keep testing, but maybe some days later. & If I encountered to any other crash, Ill report back here.


[http_downloader_settings.zip](https://github.com/erickutcher/httpdownloader/files/3924060/http_downloader_settings.zip)

**Comment by rayman3003** [comment]
Update: one more time, the new beta 2 get crashed.

I was downloading the following list, & for some reason, My internet disconnected & then suddenly, "HTTP_Downloader_64_1.0.3.0_Beta_2" get crashed in the middle of downloads. (Setting was exactly like the one I posted above).

[2.txt](https://github.com/erickutcher/httpdownloader/files/3941992/2.txt)

**Comment by erickutcher** [comment]
Try disabling the temporary download directory in the Advanced options. See if that makes any difference.

I'm not going to be able to work on this for a long while. My main computer is pretty much dead. Until I can save enough money to build a new one, this project is going to be on hold.

**Comment by rayman3003** [comment]
> 
> 
> Try disabling the temporary download directory in the Advanced options. See if that makes any difference.
> 
> I'm not going to be able to work on this for a long while. My main computer is pretty much dead. Until I can save enough money to build a new one, this project is going to be on hold.

"temporary download directory" in old "1.0.2.4" version of "HTTP Downloader" is enabled & It never ever crashed, even 1 time!

So it is nothing to do with "temporary download directory", But a bug in newer version(s).

Bad news about ur PC. Hope it gets fixed soon.

**Comment by Zero3K** [comment]
@erickutcher Have you thought about asking for donations towards getting a new PC?

**Comment by smaragdus** [comment]
@Zero3K 

There is donation button on the developer's [website](https://erickutcher.github.io/), so anyone who would like to may donate. I suppose that the developer will appreciate any donation. I believe that open source developers should be supported by anyone who can afford it.

**Comment by erickutcher** [comment]
Donations are nice, but I will never ask for them. Too much pride for that.

I'm going to scrap my current graphics card and hard drives to use in the system that I'm building. I bought a nice case and motherboard already. Just need a power supply, CPU, and RAM. I have those three chosen, but I'm waiting for the price to drop before I get them. I'm going with AMD this time. Intel is nice and all, but these new Ryzen 3000 CPUs just look too good not to get.

**Comment by rayman3003** [comment]
> 
> 
> Donations are nice, but I will never ask for them. Too much pride for that.
> 
> I'm going to scrap my current graphics card and hard drives to use in the system that I'm building. I bought a nice case and motherboard already. Just need a power supply, CPU, and RAM. I have those three chosen, but I'm waiting for the price to drop before I get them. I'm going with AMD this time. Intel is nice and all, but these new Ryzen 3000 CPUs just look too good not to get.

Yeah, Currently, The AMD CPUs have better price/performance than Intel.
For power supply, I suggest "Cooler Master".
And for "RAM", grab at least 16 GB. (But 32 suggested)

**Comment by erickutcher** [comment]
I got my spare computer set up like my old one and have been able to test the code again.

I found a stack overflow bug that could happen if you have a large queue and your network cuts out. You mentioned that the program crashed when your Internet disconnected. That might be the cause.

See if you can recreate the bug by doing the follow:

Load a bunch of links so that you end up with a large queue list. Then open up a command prompt and type: ipconfig /release

If the program crashes after the Internet is disconnected, then it's mostly like the stack overflow bug.

You can restore the Internet by typing: ipconfig /renew

**Comment by rayman3003** [comment]
> 
> 
> I got my spare computer set up like my old one and have been able to test the code again.
> 
> I found a stack overflow bug that could happen if you have a large queue and your network cuts out. You mentioned that the program crashed when your Internet disconnected. That might be the cause.
> 
> See if you can recreate the bug by doing the follow:
> 
> Load a bunch of links so that you end up with a large queue list. Then open up a command prompt and type: ipconfig /release
> 
> If the program crashes after the Internet is disconnected, then it's mostly like the stack overflow bug.
> 
> You can restore the Internet by typing: ipconfig /renew

Yes, It crashed.

**Comment by erickutcher** [comment]
Here's a new beta version for you to test. It should fix the stack overflow crash. 
[HTTP_Downloader_64_1.0.3.0_Beta_3.zip](https://github.com/erickutcher/httpdownloader/files/4028784/HTTP_Downloader_64_1.0.3.0_Beta_3.zip)

Test it out the same way you've done with the other versions. Let me know what you find.

**Comment by rayman3003** [comment]
> 
> 
> Here's a new beta version for you to test. It should fix the stack overflow crash.
> [HTTP_Downloader_64_1.0.3.0_Beta_3.zip](https://github.com/erickutcher/httpdownloader/files/4028784/HTTP_Downloader_64_1.0.3.0_Beta_3.zip)
> 
> Test it out the same way you've done with the other versions. Let me know what you find.

I tried as u said: Load a download list & used "ipconfig /release" in the middle of download to disconnect from internet. "HTTP_Downloader_64_1.0.3.0_Beta_3" didn't crashed. I reconnect to internet & removed all the download lists from "HTTP_Downloader_64_1.0.3.0_Beta_3".

And then I went to surf the internet (by browser) to find another list to test it again with "HTTP_Downloader_64_1.0.3.0_Beta_3".

While I was searching, Suddenly "HTTP_Downloader_64_1.0.3.0_Beta_3" [That had no download list in the queue] get crashed (as always without any warning!)

This is a new type of crash. Crash without any download list!

**Comment by erickutcher** [comment]
Not sure why it would crash when nothing is going on. There hasn't been any change that could cause that. Anyway, keep testing it out without manually disconnecting. Use it like you did with the first two betas.

**Comment by rayman3003** [comment]
> 
> 
> Not sure why it would crash when nothing is going on. There hasn't been any change that could cause that. Anyway, keep testing it out without manually disconnecting. Use it like you did with the first two betas.

The download list that I used (before the crash), was the one in the first post (1.txt) that has 9999 links in it. I removed the list, and after some seconds, It crashed. (without any list in download queue.)

U can use the (1.txt) in the first post. Add it to the "HTTP_Downloader_64_1.0.3.0_Beta_3" urself & then remove it, stop it, pause it, start it, restart it... & I'm sure after some tries, U will see the crash.

**Comment by erickutcher** [comment]
Time to do this hardcore mode.

Here's a debug version that's set up to generate crash dumps that I can analyze on my system. 
[HTTP_Downloader_64_1.0.3.0_Beta_3_Debug.zip](https://github.com/erickutcher/httpdownloader/files/4033544/HTTP_Downloader_64_1.0.3.0_Beta_3_Debug.zip)

It's set up to run in portable mode so it doesn't affect your current install.

First thing you'll want to do after extracting the zip is to run HTTP Downloader and set up your settings the way you like them. After you've done that, close the program. The next thing you'll want to do is run the procdump64 program that's in the same folder. It'll prompt you to accept a license agreement. Accept the license and if the program doesn't close automatically, you can close it manually. Finally, you'll want to run the DEBUG_HTTP_DOWNLOADER.bat file. What it does is runs procdump64 which then loads HTTP Downloader and waits for it to crash. Use HTTP Downloader like you normally would to get it to crash. When/if it does crash, both programs should exit and there'll be a dump file in the CRASH_DUMPS folder. Zip the dump file and upload it here so I can analyze it.

This was the easiest way I could come up with to get directly at the problem. Hopefully it'll resolve things.

**Comment by rayman3003** [comment]
> 
> 
> Time to do this hardcore mode.
> 
> Here's a debug version that's set up to generate crash dumps that I can analyze on my system.
> [HTTP_Downloader_64_1.0.3.0_Beta_3_Debug.zip](https://github.com/erickutcher/httpdownloader/files/4033544/HTTP_Downloader_64_1.0.3.0_Beta_3_Debug.zip)
> 
> It's set up to run in portable mode so it doesn't affect your current install.
> 
> First thing you'll want to do after extracting the zip is to run HTTP Downloader and set up your settings the way you like them. After you've done that, close the program. The next thing you'll want to do is run the procdump64 program that's in the same folder. It'll prompt you to accept a license agreement. Accept the license and if the program doesn't close automatically, you can close it manually. Finally, you'll want to run the DEBUG_HTTP_DOWNLOADER.bat file. What it does is runs procdump64 which then loads HTTP Downloader and waits for it to crash. Use HTTP Downloader like you normally would to get it to crash. When/if it does crash, both programs should exit and there'll be a dump file in the CRASH_DUMPS folder. Zip the dump file and upload it here so I can analyze it.
> 
> This was the easiest way I could come up with to get directly at the problem. Hopefully it'll resolve things.

Good, Good. Here is the crash dumps:

[CRASH_DUMPS.zip](https://github.com/erickutcher/httpdownloader/files/4034394/CRASH_DUMPS.zip)

**Comment by erickutcher** [comment]
I see where it's crashing, but it's going to be a difficult one to trace.

If you can get it to consistently crash the same way, then let me know the exact steps that you did to do it. I've been loading links, stopping, starting, removing, etc, but I'm not able to get it to happen.

Get a couple more crash dumps and let me know step by step what you did to get it to crash. It also helps to know how long it took to do each step. For example, how long you wait until you removed or stopped the downloads after they were added, etc.

**Comment by rayman3003** [comment]
> 
> 
> I see where it's crashing, but it's going to be a difficult one to trace.
> 
> If you can get it to consistently crash the same way, then let me know the exact steps that you did to do it. I've been loading links, stopping, starting, removing, etc, but I'm not able to get it to happen.
> 
> Get a couple more crash dumps and let me know step by step what you did to get it to crash. It also helps to know how long it took to do each step. For example, how long you wait until you removed or stopped the downloads after they were added, etc.

- I used **30** "Active Download Limit".
- I used only the list in the first post -> "1.txt". 
- It took less than 1 minute for crash to get happened.
- Maybe this is not important, But the website in the "1.txt" is blocked by my ISP, So "HTTP_Downloader_64_1.0.3.0_Beta_3_Debug" didn't download any of the links. I just stopped all the links -> Pause all the links -> restart all the links -> And boom, crashed happened.

One more time I say: I did download all these links & more (more than 40000 links) some weeks ago with "HTTP Downloader **1.0.2.4**" [used VPN], without a single crash!

**Comment by erickutcher** [comment]
I was able to confirm that the stack overflow crash happens in 1.0.2.4. An unstable connection would have been what could cause it.

Anyway, you'll probably have to record a video for me showing everything you do to cause the crash because I simply can't create it on my end. I've done numerous pauses, stops, restarts, removes, etc. and it's just not crashing.

Version 1.0.2.7 is the first version that I implemented FTP support. If the crash happens in that version and not in 1.0.2.6, then I could narrow down the code that might be affected, but that's all I can think to do at the moment.

**Comment by rayman3003** [comment]
> 
> 
> I was able to confirm that the stack overflow crash happens in 1.0.2.4. An unstable connection would have been what could cause it.
> 
> Anyway, you'll probably have to record a video for me showing everything you do to cause the crash because I simply can't create it on my end. I've done numerous pauses, stops, restarts, removes, etc. and it's just not crashing.
> 
> Version 1.0.2.7 is the first version that I implemented FTP support. If the crash happens in that version and not in 1.0.2.6, then I could narrow down the code that might be affected, but that's all I can think to do at the moment.

OK, Here is a 1 minutes video that shows how fast "HTTP_Downloader_64_1.0.3.0_Beta_3_Debug" getting crashed:

https://gofile.io/?c=DXowIY

or

https://waaw.tv/watch_video.php?v=6AnvvCiolEAf

**Comment by erickutcher** [comment]
Excellent. Now that I know the states that the downloads enter, I should be able to recreate it. I just need to find a site that'll stall after it's added. My ISP doesn't block that particular site so I'll have to go hunt for a similarly behaving one.

**Comment by rayman3003** [comment]
> 
> 
> Excellent. Now that I know the states that the downloads enter, I should be able to recreate it. I just need to find a site that'll stall after it's added. My ISP doesn't block that particular site so I'll have to go hunt for a similarly behaving one.

Maybe & maybe if u block "HTTP Downloader" in your firewall, It gives the similar behavior.
I use this free good firewall:
[https://github.com/henrypp/simplewall](url)

**Comment by erickutcher** [comment]
Already tried it and it just causes it to time out. What's happening on your end is that a connection is established, but it gets stalled. Either the HTTP request never gets sent, or a response is never received. That's why it shows that infinity symbol. I've come across sites that have done that before, so it's just a matter of finding one again to simulate your environment.

**Comment by rayman3003** [comment]
> 
> 
> Already tried it and it just causes it to time out. What's happening on your end is that a connection is established, but it gets stalled. Either the HTTP request never gets sent, or a response is never received. That's why it shows that infinity symbol. I've come across sites that have done that before, so it's just a matter of finding one again to simulate your environment.

OK, Take ur time.
And if u need any other test, Just tell me.

**Comment by erickutcher** [comment]
See if you can get it to crash the same way you did in the video, but without any queued items in the list. So if your active download limit is set to 30, then just load 30 URLs. If it crashes with 30, then see if it'll do it with just 1.

The bug you're experiencing is related to those connecting downloads not getting stopped. They're getting stuck with that infinity symbol and should have transitioned into the stop state (where the connection gets cleaned up). Because it's not getting cleaned up, when you restart the download and then stop it again, it tries to clean up the memory twice and that's what's causing the crash.

I just need to figure out why and where it's getting stuck and how to get it unstuck.

**Comment by rayman3003** [comment]
> 
> 
> See if you can get it to crash the same way you did in the video, but without any queued items in the list. So if your active download limit is set to 30, then just load 30 URLs. If it crashes with 30, then see if it'll do it with just 1.
> 
> The bug you're experiencing is related to those connecting downloads not getting stopped. They're getting stuck with that infinity symbol and should have transitioned into the stop state (where the connection gets cleaned up). Because it's not getting cleaned up, when you restart the download and then stop it again, it tries to clean up the memory twice and that's what's causing the crash.
> 
> I just need to figure out why and where it's getting stuck and how to get it unstuck.
> 
> Edit:
> 
> Here's the last version that I released before I added FTP support (version 1.0.2.6): https://github.com/erickutcher/httpdownloader/releases/download/v1.0.2.6/HTTP_Downloader_64.zip
> 
> Here's the first one with FTP support (version 1.0.2.7): https://github.com/erickutcher/httpdownloader/releases/download/v1.0.2.7/HTTP_Downloader_64.zip
> 
> If you're able to test those two versions using the same URLs and perform the exact same steps that you did in the video, then that'll help me narrow things down further. If it wasn't doing it in earlier versions, like you mentioned, then it has to be the changes I made between those two versions.

1- As u requested, This time, I loaded only 30 links from 1.txt into "1.0.3.0_Beta_3_Debug". And.... It crashed. But this time, It took a later more to crash. I stopped, paused, restart for 3-4 times, each, & then crash happened.

2- Did the same test with "1.0.2.7". It did crashed too. But lasted more. It crashed after 7-8 times that I tried (stop, pause, restart, start).

3- One more time, I got back to "1.0.2.6" & did the same test with it (all 9999 links), & nothing. It is the most stable downloader in the world! I tried (stop, pause, restart, start) more than 30 times & it still stays there without a single blink!

These tests show that maybe the FTP is the root of these crashes, But I'm not sure. Bcuz the last version "1.0.3.0" (that has more options than 1.0.2.7) crashed sooner than 1.0.2.7.

**Comment by erickutcher** [comment]
The FTP code is fairly well separated from the HTTP stuff so it shouldn't have affected anything. I did revert one bit of code that I introduced in1.0.2.7 that _could_ be the issue, but I'm not sure. We'll see.

Here's a new beta: [HTTP_Downloader_1.0.3.0_Beta_4_Debug.zip](https://github.com/erickutcher/httpdownloader/files/4049902/HTTP_Downloader_1.0.3.0_Beta_4_Debug.zip)

There's three versions of HTTP Downloader in that zip file.

HTTP_Downloader_a.exe will display a message box when an HTTPS connection is established and a request is ready to be sent. Run that one with just one link from the list you created. For example: `https://www.vgmusic.com/new-files/Illusion_of_Gaia_DungeonTheme.mid` If it's able to connect properly, then there should be a message box saying "HTTPS established". Let me know if that message box shows up. You don't need to test if this version will crash.

HTTP_Downloader_b.exe has the reverted code in it. Test this like the other versions and see if it crashes.

HTTP_Downloader_c.exe has the reverted code and a different (safer) implementation in which restarted downloads are handled. Test this one like the others to see if it crashes.

**Comment by rayman3003** [comment]
> at one with

- "HTTPS established" message didn't show up for "HTTP_Downloader_a.exe".

- I tried the the "9999 links test" with "HTTP_Downloader_b.exe" & "HTTP_Downloader_c.exe" & fortunately both passed the test successfully, without a single crach! Looks like the crach problem get solved. Thank u.

**Comment by erickutcher** [comment]
The message not showing up in version 'a' tells me that your ISP is blocking the SSL/TLS handshake. There isn't anything that I can do to fix that. A VPN (like you're using) or proxy will get around the problem.

The reverted change that I made in version 'b' and 'c' was so minor that I'm surprised it had an affect. It wasn't an obvious bug either since it happens as a result of a race condition. The only way to get a grasp of what's going is to imagine every possibility in your head. Takes a while to finally see it. It's good to know it's working again.

**Comment by erickutcher** [comment]
Latest version is up.
